### ISSUE TITLE
How does FastMoE compare to the Switch Transformer

### ISSUE THREAD

**Issue by Tiiiger** [laekov/fastmoe#25]
Hi there, 

Thank you for releasing this repo! This is a fantastic system and will benefit the research of many of us.

Quick question: how does FastMoE compare to the Switch Transformer? 

I understand that a direct comparison is not feasible because mesh-tensorflow requires TPUs. But do you have some approximate metrics?

For example, by looking at Figure 7, I see that applying FastMoE only save roughly 30% of wall clock time in order to reach the final loss of the baseline. However, my impression of reading the Switch transformer paper is that applying MoE is 6-7x faster when training a T5-base model. 

I understand that there are several variables that are not controlled here. Hardware, models, training hyperparameters are all different. That said, I am curious to hear your perspective on this. What are the key factors that make FastMoE much slower?

Thank you in advance!

**Comment by laekov** [comment]
FastMoE is a training system, while Switch Transformer is a model. You can implement a equivalent switch transformer model with the functions provided by FastMoE (some of them like routing strategy are still being developed). 

For performance metric (training speed), I think the only metric that can bridge different hardware is utilization, i.e. training throughput / peak throughput of the cluster. Unfortunately, I do not see the switch tansformer team report their utilization. 

For figure 7, the _T5 is 6-7x faster_ is compared __on per step basis__, which equals to figure 7 left. We did not calculate the actual speed up because we do not want to fix the loss value which means the model is okay. You may find out the speedup from figure 7 left, which should be similar to the result in switch transformer.

**Comment by Tiiiger** [comment]
hi @laekov ,

Thank you for the answer! Should have made the question clearer: how do you compare the MoE model implemented in the FastMoE paper to switch transformer implemented in mesh-tensorflow?

Are you suggesting that switch transformer (considering both the model and the system) because it has a better routing strategy?

I got th 6-7x number from the figure 5 in the switch transformer paper, where the x-axis is time, not steps.

**Comment by laekov** [comment]
> hi @laekov ,
> 
> Thank you for the answer! Should have made the question clearer: how do you compare the MoE model implemented in the FastMoE paper to switch transformer implemented in mesh-tensorflow?
> 
> Are you suggesting that switch transformer (considering both the model and the system) because it has a better routing strategy?
> 
> I got th 6-7x number from the figure 5 in the switch transformer paper, where the x-axis is time, not steps.

FastMoE is of course not that strong as switch transformer and mtf, as we are only beginning to implement the system that can train an MoE model. 

We mentioned the model in our arxiv article to show that FastMoE can train real models, but we do not suggest that either the model or the system is stronger or comparable with Google's. We do hope that FastMoE can be stronger in the future, but currently, we have neither strong human resource nor abundant hardware resource. 

I was looking at figure 8 in switch's article when I said training step. You are right that switch transformer shows 7x speedup in training time against its baseline.

**Comment by Tiiiger** [comment]
Got it! thank you : )
